worker/src/primary_connector.rs

```diff
@@ -3,94 +3,135 @@
                 Some(digest) = self.rx_digest.recv() => {
-                    // We don't care about the error
-                    let _ = self.primary_client.send(&digest).await;
+                    let address = self.committee
+                        .primary(&self.name)
+                        .expect("Our public key is not in the committee")
+                        .worker_to_primary;
+                    _handle = self.primary_client.send(address, &digest).await;
                 },
+
                 // Trigger reconfigure.
                 result = self.rx_reconfigure.changed() => {
                     result.expect("Committee channel dropped");
                     let message = self.rx_reconfigure.borrow().clone();
-                    if let Reconfigure::Shutdown(_token) = message {
-                        return;
+                    match message {
+                        ReconfigureNotification::NewCommittee(new_committee) => {
+                            self.committee = new_committee;
+                        },
+                        ReconfigureNotification::Shutdown => return
                     }
                 }
             }
         }
     }
 }
 
-#[derive(Clone)]
-pub struct PrimaryClient {
-    /// The primary network address.
-    _primary_address: Multiaddr,
-    client: WorkerToPrimaryClient<Channel>,
+pub struct WorkerToPrimaryNetwork {
+    client: Option<WorkerToPrimaryClient<Channel>>,
+    config: mysten_network::config::Config,
+    retry_config: RetryConfig,
+    executor: BoundedExecutor,
 }
 
-impl PrimaryClient {
-    pub fn new(address: Multiaddr) -> Self {
-        //TODO don't panic here if address isn't supported
-        let config = mysten_network::config::Config::new();
-        let channel = config.connect_lazy(&address).unwrap();
-        let client = WorkerToPrimaryClient::new(channel);
+impl Default for WorkerToPrimaryNetwork {
+    fn default() -> Self {
+        let retry_config = RetryConfig {
+            // Retry for forever
+            retrying_max_elapsed_time: None,
+            ..Default::default()
+        };
 
         Self {
-            _primary_address: address,
-            client,
+            client: Default::default(),
+            config: Default::default(),
+            retry_config,
+            executor: BoundedExecutor::new(MAX_TASK_CONCURRENCY, Handle::current()),
         }
     }
+}
 
-    pub async fn send(&mut self, message: &WorkerPrimaryMessage) -> anyhow::Result<()> {
-        let message = BincodeEncodedPayload::try_from(message)?;
-        self.client
-            .send_message(message)
-            .await
-            .map_err(Into::into)
-            .map(|_| ())
+impl WorkerToPrimaryNetwork {
+    pub async fn send<PublicKey: VerifyingKey>(
+        &mut self,
+        address: Multiaddr,
+        message: &WorkerPrimaryMessage<PublicKey>,
+    ) -> CancelHandler<()> {
+        if self.client.is_none() {
```

Comment:
I think it's quite implicit the initialisation of the client here. This method once the client is created it will not initialise a new one even if we need to communicate to a different address. I am aware that we don't expect the primary's address to change, but since this is the case I would refactor to initialise the client once the `WorkerToPrimaryNetwork` is created. This will make clear that when we `send` a message it will always be sent to the right address.

Otherwise, if we expect the primary's address to change via reconfiguration - then again this should be refactored as the client will keep pointing to the old address

Reply:
Fixed